### PR TITLE
feat: autocomplete endpoint for elasticsearch

### PR DESCRIPTION
**Pull request by CaiWillis**
TIS21-3279: Create endpoint that returns list of assigned programme names from ES

### CODE REVIEW

**Comment by jayanta2018** [comment]
Just wondering some tests failure is there?

**Comment by CaiWillis** [comment]
Have ignored code smell regarding switch statement as it's more appropriate if in future we autocomplete other fields (e.g. TIS status, GMC outcome, Doctor name etc.)

File: application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationElasticSearchService.java [hunk 1]
```diff
@@ -82,6 +84,15 @@ public String formatDesignatedBodyCodesForElasticsearchQuery(
     return String.join(",", escapedCodes);
   }
 
+  public List<String> getAutocompleteResults(String fieldname, String input, List<String> dbcs) {
+    var results = recommendationElasticSearchRepository
+        .autocomplete(fieldname, input, formatDesignatedBodyCodesForElasticsearchQuery(dbcs));
+    return results.stream()
+        .filter(result -> Objects.nonNull(getAutocompleteFieldValue(fieldname, result)))
+        .map(result -> getAutocompleteFieldValue(fieldname, result))
+        .collect(Collectors.toList()).stream().distinct().collect(Collectors.toList());
```
Comment (JosephKelly) [inline application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationElasticSearchService.java:93]:
```java
    return results.stream()
        .map(result -> getAutocompleteFieldValue(fieldname, result))
        .filter(Objects::nonNull)
        .distinct().collect(Collectors.toList());
```

File: application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationElasticSearchService.java [hunk 2]
```diff
@@ -120,4 +131,14 @@ private Iterable<RecommendationView> findRecommendationViewsByGmcNumber(
     }
     return result;
   }
+
+  private String getAutocompleteFieldValue(String fieldName, RecommendationView result) {
```
Comment (JosephKelly) [inline application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationElasticSearchService.java:135]: Just IMO, this function is much more generic.  `getFieldValueAsString` seems like a more fitting description.

File: application/src/test/java/uk/nhs/hee/tis/revalidation/controller/DoctorsForDBControllerTest.java [hunk 3]
```diff
@@ -272,6 +280,38 @@ void shouldReturnDoctorsByGmcId() throws Exception {
...
+    this.mockMvc.perform(get(GET_AUTOCOMPLETE)
+            .param(AUTOCOMPLETE_FIELD, fieldNameParam)
+            .param(INPUT, inputParam)
+            .param(DESIGNATED_BODY_CODES, new String[]{"1-AIIDWI"}))
+        .andExpect(status().isOk())
+        .andExpect(content().json(mapper.writeValueAsString(result)));
+  }
+
+  @Test
+  void shouldRejectAutocompleteResultsForIvalidField() throws Exception {
```
Comment (JosephKelly) [inline application/src/test/java/uk/nhs/hee/tis/revalidation/controller/DoctorsForDBControllerTest.java:309]: Typo: `Invalid`

Comment (JosephKelly) [inline application/src/test/java/uk/nhs/hee/tis/revalidation/controller/DoctorsForDBControllerTest.java:309]: Misleading name suggests that the problem is the `fieldName` param... the absence of input and/or dbcs causes the HTTP 400.

`@RequestParam(required = true, ...)` is the default

**Review by jayanta2018 — APPROVED** [review]

File: application/src/test/java/uk/nhs/hee/tis/revalidation/JsonSerializationTest.java [hunk 4]
```diff
@@ -36,11 +36,15 @@
...
 import uk.nhs.hee.tis.revalidation.dto.DoctorsForDbDto;
 import uk.nhs.hee.tis.revalidation.service.DoctorsForDBService;
+import uk.nhs.hee.tis.revalidation.service.RecommendationElasticSearchService;
 
 @ExtendWith(MockitoExtension.class)
 @WebMvcTest(DoctorsForDBController.class)
 class JsonSerializationTest {
 
+  @MockBean
+  private RecommendationElasticSearchService recommendationElasticSearchService;
```
Comment (JosephKelly) [inline application/src/test/java/uk/nhs/hee/tis/revalidation/JsonSerializationTest.java:46]: I know this PR doesn't create this issue but this test should not be a `@WebMvcTest`.  I'm assuming the use of `@WebMvcTest(DoctorsForDBController.class)` is the only reason that you've needed to Mock the new dependency here.

File: application/src/main/java/uk/nhs/hee/tis/revalidation/repository/RecommendationElasticSearchRepository.java [hunk 5]
```diff
@@ -45,6 +45,9 @@ Page<RecommendationView> findByUnderNotice(final String searchQuery,
   Page<RecommendationView> findAll(final String searchQuery,
       final String dbcs, List<String> hiddenGmcIds, final Pageable pageable);
 
+  @Query("{\"bool\":{\"filter\":[{\"match\":{\"underNotice\":\"YES\"}},{\"match\":{\"designatedBody\":\"?2\"}},{\"match_phrase_prefix\":{\"?0\":{\"query\":\"?1\"}}}]}}")
+  List<RecommendationView> autocomplete(final String fieldName, final String input, final String dbcs);
```
Comment (JosephKelly) [inline application/src/main/java/uk/nhs/hee/tis/revalidation/repository/RecommendationElasticSearchRepository.java:49]: method name should reflect what it does,
 e.g. `findByFieldNameParameter(String, String, String)`?

File: application/src/main/java/uk/nhs/hee/tis/revalidation/controller/DoctorsForDBController.java [hunk 6]
```diff
@@ -167,6 +179,19 @@ public ResponseEntity<TraineeSummaryDto> getDoctors(
     return ResponseEntity.ok().body(TraineeSummaryDto.builder().build());
   }
 
+  @ApiOperation(value = "Get autocomplete value for field", notes = "It will return matching field values", response = ResponseEntity.class)
+  @ApiResponses(value = {
+      @ApiResponse(code = 200, message = "Suggested field values", response = ResponseEntity.class)})
+  @GetMapping(value = {"/autocomplete"})
+  public ResponseEntity<List<String>> getAutocompleteProgrammeName(
```
Comment (JosephKelly) [inline application/src/main/java/uk/nhs/hee/tis/revalidation/controller/DoctorsForDBController.java:186]: The method name should be more generic.

**Review by JosephKelly — CHANGES_REQUESTED** [review]

File: application/src/test/java/uk/nhs/hee/tis/revalidation/controller/DoctorsForDBControllerTest.java (hunk 3 above)
Comment (CaiWillis) [reply application/src/test/java/uk/nhs/hee/tis/revalidation/controller/DoctorsForDBControllerTest.java:309]: Ah this was from an old implementation that didn't work anyway, thanks for catching!

File: application/src/main/java/uk/nhs/hee/tis/revalidation/controller/DoctorsForDBController.java [hunk 7]
```diff
@@ -160,13 +160,25 @@ public ResponseEntity<TraineeSummaryDto> getDoctors(
...
     return ResponseEntity.ok().body(TraineeSummaryDto.builder().build());
   }
 
+  @ApiOperation(value = "Get autocomplete value for field", notes = "It will return matching field values", response = ResponseEntity.class)
+  @ApiResponses(value = {
+      @ApiResponse(code = 200, message = "Suggested field values", response = ResponseEntity.class)})
+  @GetMapping("/autocomplete")
+  public ResponseEntity<List<String>> getAutocompleteValesForField(
+      @NonNull @RequestParam(name = AUTOCOMPLETE_FIELD) final String fieldName,
+      @RequestParam(name = INPUT) final String input,
```
Comment (JosephKelly) [inline application/src/main/java/uk/nhs/hee/tis/revalidation/controller/DoctorsForDBController.java:175]: Did this need to be defined as `requied = false`?

File: application/src/test/java/uk/nhs/hee/tis/revalidation/JsonSerializationTest.java (hunk 4 above)
Comment (JosephKelly) [reply application/src/test/java/uk/nhs/hee/tis/revalidation/JsonSerializationTest.java:46]: I'll get this in a later PR

File: application/src/main/java/uk/nhs/hee/tis/revalidation/controller/DoctorsForDBController.java (hunk 7 above)
Comment (CaiWillis) [reply application/src/main/java/uk/nhs/hee/tis/revalidation/controller/DoctorsForDBController.java:175]: makes sense to me

**Review by JosephKelly — APPROVED** [review]

**Pull request merged**